plugins/amdgpu/amdgpu_plugin_topology.c

```diff
@@ -1098,6 +1135,63 @@ bool match_xgmi_groups(struct tp_system *src_sys, struct tp_system *dest_sys,
 	return false;
 }
 
+/* return 1  	 if gpu override is set
+ * return 0  	 if no gpu overide
+ * return -errno if failed to set gpu override */
+int get_user_gpu_override(struct tp_system *src_sys, struct tp_system *dest_sys,
```

Comment:
If I understand this correctly, it applies very different rules for the override. For example it doesn't do any XGMI hive matching. And it doesn't have any flexibility in remapping source GPUs to a compatible destination GPU. It has to match in the order specified.

Maybe it would be easier, more flexible and more correct to use KFD_DESTINATION_GPUS as a filter that specifies which GPUs are available for mapping. I.e. just remove the GPUs not specified from dest_sys. Then you can use the normal match_xgmi_groups function to do the actual matching, but with a subset of the GPUs.

Reply:
The intention of this parameter was to allow the user to force the GPUs to be mapped in a certain order. So for example if the GPUs were A1, A2, A3, A4 on the source node and the mapping code would map them to B1, B2, B3, B4, but for unforeseen reasons, the user would want the GPUs to be mapped to B3, B4, B2, B1.
With current code, the number of GPUs on destination node needs to match number of GPUs on source node. If the destination node has more GPUs, the user is expected to mask the extra GPUs using cgroups.

The check here:
```
/* Ignore extra GPU's */
if (index >= dest_sys->num_nodes)
    break;
    
```
Is in case a user specified: KFD_DESTINATION_GPUS=0x0001, 0x0002, 0x0003, 0x0004 (for e.g because they have a generic script for all their systems), but there are only 3 GPUs in the system.

Reply:
That's weird. The only way that you can have index >= dest_sys->num_nodes is, if some of the GPU names are invalid. So shouldn't that be an error?

Another prossible error is, if the same GPU is specified multiple times. Is there a way to catch such an error?

Masking GPUs with cgroups won't work for MS running in a privileged docker container. So that means they'll be forced to use KFD_DESTINATION_GPUS. We need to make this as robust as possible (i.e. catch as many user errors as possible). That includes:

- bad GPU IDs
- same GPU specified multiple times
- specified list of destination GPUs incomatible with the source link topology and XGMI topology

Duplicating all the link compatibility checking in the override function is error prone and harder to maintain. Therefore I would strongly recommend, relying on as much common code as possible. If that means letting the common code reorder the destination GPUs, I think that's OK.

Reply:
To be honest, being able to re-order the GPU's was the main reason why I implemented this variable (I used it mostly for internal testing/debugging) and it was meant to be more of a power-user feature, which is why I did not implemented too many validation checks. But I can see the use case for the way you want to use it and also how the environment variable name could be misleading. Would you be ok if I implement this:

KFD_DESTINATION_GPUS: List of allowed GPUs on restore node. GPUs in this list may be re-ordered.
KFD_OVERRIDE_GPUS: Will be used to force a specific GPU order.

For both cases, I will add checks for invalid and duplicate GPU ID's and/or minor render numbers.

Note: Current code will catch case where destination GPU does not have same XGMI topology as source GPU. In map_device function we go through list of iolinks and call iolink_match() function to compare them.

Reply:
I removed this feature :)